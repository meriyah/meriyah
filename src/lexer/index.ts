export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
