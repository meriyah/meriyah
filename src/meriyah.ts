import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
