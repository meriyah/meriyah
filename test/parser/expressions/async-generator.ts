import * as t from 'node:assert/strict';
import { outdent } from 'outdent';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail, pass } from '../../test-utils';

describe('Expressions - Async Generator', () => {
  for (const arg of [
    '(async function *foo() { var await; });',
    '(async function *foo() { void await; });',
    '(async function *foo() { await: ; });',
    '(async function *foo(x = 1) {"use strict"})',
    '(async function *foo(foo) { super() })',
    '(async function *foo(foo) { super.prop });',
    '(async function *foo(foo = super()) { var bar; });',
    '(async function*([...x, y]) {})',
    '(async function*([...x, y] = [1, 2, 3]) {})',
    '(async function* h([...{ x } = []]) {})',
    '(async function* h([...{ x } = []] = []) {})',
    '(async function*(x = await 1) { });',
    '(async function*() { await: 1; });',
    '(async function *foo(...a,) {})',
    '(async function *foo([...[ x ] = []])',
    '(async function *foo([...{ x } = []]) {})',
    '(async function *foo([...{ x } = []] = []) {})',
    '(async function *foo([...x, y]) {})',
    '(async function *foo([...x = []] = []) {})',
    '(async function *foo(...a,) {})',
    '(async function *foo([...[x], y] = [1, 2, 3]) {})',
    '(async function *foo([...{ x }, y] = [1, 2, 3])',
    '(async function *foo([...{ x }, y])',
    '(async function *foo([...{ x } = []] = [])',
    '(async function *foo([...{ x } = []])',
  ]) {
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`);
      });
    });

    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { lexical: true });
      });
    });

    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { next: true });
      });
    });

    it(`"use strict"; ${arg}`, () => {
      t.throws(() => {
        parseSource(`"use strict"; ${arg}`);
      });
    });

    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { sourceType: 'module' });
      });
    });

    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { webcompat: true });
      });
    });
  }

  for (const arg of [
    '(async function *foo() { }.prototype)',
    '(async function *foo(x, y = x, z = y) { })',
    '(async function *foo(x = y, y) { })',
    '(async function *foo(a, b = 39,) { })',
    '(async function *foo(a, b,) { })',
    '(async function *foo(_ = (function() {}())) { })',
    '(async function *([x = 23]) { })',
    '(async function *([{ x }]) { })',
    '(async function *(x = x) { })',
    '(async function*([...[...x]]) { })',
    '(async function *foo([...x] = 123) { })',
    '(async function *foo({ cls = class {}, xCls = class X {}, xCls2 = class { static name() {} } } = {}) {})',
    '(async function*({ w: [x, y, z] = [4, 5, 6] } = { w: [7, undefined, ] }) { })',
    '(async function*({ w: { x, y, z } = { x: 4, y: 5, z: 6 } } = { w: undefined }) { })',
    '(async function* h([[,] = g()]) { })',
    '(async function* g([[x]]) { })',
    '(async function* h([cls = class {}, xCls = class X {}, xCls2 = class { static name() {} }]) { })',
    '(async function* h([fn = function () {}, xFn = function x() {}]) { })',
    '(async function* h([{ x, y, z } = { x: 44, y: 55, z: 66 }]) { })',
    '(async function* h([]) { })',
    '(async function* h([...[,]]) { })',
    '(async function* g([...x]) { })',
    '(async function* h([fn = function () {}, xFn = function x() {}] = []) { })',
    '(async function* h([x] = []) { })',
    '(async function* h({} = null) { })',
    'var gen = async function *() { yield [...yield]; };',
    '(async function* h({a, b, ...rest} = {x: 1, y: 2, a: 5, b: 3}) { })',
    '(async function* h({ x, }) { })',
    '(async function* h({ w: [x, y, z] = [4, 5, 6] }) { })',
    '(async function*({}) { })',
    '(async function*({ x, }) { })',
    '(async function*({ x: y = 33 }) { })',
    outdent`
      var gen = async function *g() {
        yield [...yield];
      };
    `,
    outdent`
      var gen = async function *() {
        yield {
            ...yield yield,
            ...(function(arg) {
              var yield = arg;
              return {...yield};
            }(yield)),
            ...yield,
          }
      };
    `,
    outdent`
      var gen = async function *g() {
        return (function(arg) {
            var yield = arg + 1;
            return yield;
          }(yield))
      };
    `,
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`);
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { lexical: true });
      });
    });
    it(`() => { ${arg} }`, () => {
      t.doesNotThrow(() => {
        parseSource(`() => { ${arg} }`);
      });
    });

    it(`function foo() { ${arg}}`, () => {
      t.doesNotThrow(() => {
        parseSource(`function foo() { ${arg}}`);
      });
    });

    it(`function foo() { ${arg}}`, () => {
      t.doesNotThrow(() => {
        parseSource(`function foo() { ${arg}}`, { next: true });
      });
    });
  }
  fail('Expressions - Async Generator (pass)', [
    '(async function*(a = super()) { });',
    '0, async function* g(...a,) {};',
    '(async function* yield() { });',
    '(async function* g() { var await; });',
    outdent`
      "use strict";
        async function *g() {
        return {
             ...(function() {
                var yield;
             }()),
          }
      };
    `,
    '(async function* g() { var yield; });',
    '(async function*(a = super()) { });',
    '(async function*() { } = 1);',
    '(async function *() { var await; })',
    '(async function*([...x, y] = [1, 2, 3]) {})',
    '(async function* h([...{ x } = []]) {})',
    '(async.foo6 => 1)',
    '(async function* h([...{ x } = []] = []) {})',
    '(async function*(x = await 1) { });',
    '(async function*() { await: 1; });',
    '(async function *foo(...a,) {})',
    '(async function *foo([...[ x ] = []])',
    '(async function *foo([...{ x } = []]) {})',
    '(async function *foo([...{ x } = []] = []) {})',
    '(async function *foo([...x, y]) {})',
    '(async function *foo([...x = []] = []) {})',
    '(async function *foo(...a,) {})',
    '(async function *foo([...[x], y] = [1, 2, 3]) {})',
    '(async function *foo([...{ x }, y] = [1, 2, 3])',
    '(async function *foo([...{ x }, y])',
    '(async function *foo([...{ x } = []] = [])',
    '(async function *foo([...{ x } = []])',
    '(async function* yield() { });',
    '(async function* g() { var await; });',
    { code: '(async function* g() { void await; });', options: { sourceType: 'module' } },
    '(async function* g() { void yield; });',
    '0, async function* g(...x = []) {}',
    '(async function *foo([...{ x } = []])',
    '(async function *foo([...{ x } = []])',
    '(async function *foo([...{ x } = []])',
    '(async function *foo([...{ x } = []])',
  ]);

  pass('Expressions - Async Generator (pass)', [
    '(async function* h([cls = class {}, xCls = class X {}, xCls2 = class { static name() {} }]) { })',
    '(async function* h([fn = function () {}, xFn = function x() {}] = []) { })',
    '(async function* h([{ x, y, z } = { x: 44, y: 55, z: 66 }]) { })',
    '(async function *([{ x }]) { })',
    '(async function*({ w: { x, y, z } = { x: 4, y: 5, z: 6 } } = { w: undefined }) { })',
    '(async function *foo() { }.prototype)',
    '(async function *foo([...x] = 123) { })',
    '(async function *foo(x, y = x, z = y) { })',

    '(async function* h([]) { })',
  ]);
});
