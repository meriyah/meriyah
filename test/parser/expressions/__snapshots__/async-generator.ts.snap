// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > "use strict";
  async function *g() {
  return {
       ...(function() {
          var yield;
       }()),
    }
}; 1`] = `
"SyntaxError [5:14-5:19]: Unexpected strict mode reserved word
  3 |   return {
  4 |        ...(function() {
> 5 |           var yield;
    |               ^^^^^ Unexpected strict mode reserved word
  6 |        }()),
  7 |     }
  8 | };"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *() { var await; }) 1`] = `
"SyntaxError [1:26-1:31]: cannot use "await" as identifier inside an async function
> 1 | (async function *() { var await; })
    |                           ^^^^^ cannot use "await" as identifier inside an async function"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *([{ x }]) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": null,
        "params": [
          {
            "elements": [
              {
                "properties": [
                  {
                    "computed": false,
                    "key": {
                      "name": "x",
                      "type": "Identifier",
                    },
                    "kind": "init",
                    "method": false,
                    "shorthand": true,
                    "type": "Property",
                    "value": {
                      "name": "x",
                      "type": "Identifier",
                    },
                  },
                ],
                "type": "ObjectPattern",
              },
            ],
            "type": "ArrayPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo() { }.prototype) 1`] = `
{
  "body": [
    {
      "expression": {
        "computed": false,
        "object": {
          "async": true,
          "body": {
            "body": [],
            "type": "BlockStatement",
          },
          "generator": true,
          "id": {
            "name": "foo",
            "type": "Identifier",
          },
          "params": [],
          "type": "FunctionExpression",
        },
        "optional": false,
        "property": {
          "name": "prototype",
          "type": "Identifier",
        },
        "type": "MemberExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo(...a,) {}) 1`] = `
"SyntaxError [1:25-1:26]: The binding declaration is not destructible
> 1 | (async function *foo(...a,) {})
    |                          ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo(...a,) {}) 2`] = `
"SyntaxError [1:25-1:26]: The binding declaration is not destructible
> 1 | (async function *foo(...a,) {})
    |                          ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...[ x ] = []]) 1`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...[ x ] = []])
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...[x], y] = [1, 2, 3]) {}) 1`] = `
"SyntaxError [1:33-1:34]: The binding declaration is not destructible
> 1 | (async function *foo([...[x], y] = [1, 2, 3]) {})
    |                                  ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []] = []) {}) 1`] = `
"SyntaxError [1:37-1:38]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []] = []) {})
    |                                      ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []] = []) 1`] = `
"SyntaxError [1:37-1:38]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []] = [])
    |                                      ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []]) {}) 1`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []]) {})
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []]) 1`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []])
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []]) 2`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []])
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []]) 3`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []])
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []]) 4`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []])
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x } = []]) 5`] = `
"SyntaxError [1:36-1:37]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x } = []])
    |                                     ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x }, y] = [1, 2, 3]) 1`] = `
"SyntaxError [1:35-1:36]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x }, y] = [1, 2, 3])
    |                                    ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...{ x }, y]) 1`] = `
"SyntaxError [1:34-1:35]: The binding declaration is not destructible
> 1 | (async function *foo([...{ x }, y])
    |                                   ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...x = []] = []) {}) 1`] = `
"SyntaxError [1:33-1:34]: The binding declaration is not destructible
> 1 | (async function *foo([...x = []] = []) {})
    |                                  ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...x, y]) {}) 1`] = `
"SyntaxError [1:30-1:31]: The binding declaration is not destructible
> 1 | (async function *foo([...x, y]) {})
    |                               ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo([...x] = 123) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": {
          "name": "foo",
          "type": "Identifier",
        },
        "params": [
          {
            "left": {
              "elements": [
                {
                  "argument": {
                    "name": "x",
                    "type": "Identifier",
                  },
                  "type": "RestElement",
                },
              ],
              "type": "ArrayPattern",
            },
            "right": {
              "type": "Literal",
              "value": 123,
            },
            "type": "AssignmentPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function *foo(x, y = x, z = y) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": {
          "name": "foo",
          "type": "Identifier",
        },
        "params": [
          {
            "name": "x",
            "type": "Identifier",
          },
          {
            "left": {
              "name": "y",
              "type": "Identifier",
            },
            "right": {
              "name": "x",
              "type": "Identifier",
            },
            "type": "AssignmentPattern",
          },
          {
            "left": {
              "name": "z",
              "type": "Identifier",
            },
            "right": {
              "name": "y",
              "type": "Identifier",
            },
            "type": "AssignmentPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* g() { var await; }); 1`] = `
"SyntaxError [1:27-1:32]: cannot use "await" as identifier inside an async function
> 1 | (async function* g() { var await; });
    |                            ^^^^^ cannot use "await" as identifier inside an async function"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* g() { var await; }); 2`] = `
"SyntaxError [1:27-1:32]: cannot use "await" as identifier inside an async function
> 1 | (async function* g() { var await; });
    |                            ^^^^^ cannot use "await" as identifier inside an async function"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* g() { var yield; }); 1`] = `
"SyntaxError [1:27-1:32]: Yield expression not allowed in formal parameter
> 1 | (async function* g() { var yield; });
    |                            ^^^^^ Yield expression not allowed in formal parameter"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* g() { void await; }); 1`] = `
"SyntaxError [1:28-1:33]: cannot use "await" as identifier inside an async function
> 1 | (async function* g() { void await; });
    |                             ^^^^^ cannot use "await" as identifier inside an async function"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* g() { void yield; }); 1`] = `
"SyntaxError [1:33-1:34]: Invalid left-hand side in assignment
> 1 | (async function* g() { void yield; });
    |                                  ^ Invalid left-hand side in assignment"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* h([...{ x } = []] = []) {}) 1`] = `
"SyntaxError [1:35-1:36]: The binding declaration is not destructible
> 1 | (async function* h([...{ x } = []] = []) {})
    |                                    ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* h([...{ x } = []]) {}) 1`] = `
"SyntaxError [1:34-1:35]: The binding declaration is not destructible
> 1 | (async function* h([...{ x } = []]) {})
    |                                   ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* h([]) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": {
          "name": "h",
          "type": "Identifier",
        },
        "params": [
          {
            "elements": [],
            "type": "ArrayPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* h([{ x, y, z } = { x: 44, y: 55, z: 66 }]) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": {
          "name": "h",
          "type": "Identifier",
        },
        "params": [
          {
            "elements": [
              {
                "left": {
                  "properties": [
                    {
                      "computed": false,
                      "key": {
                        "name": "x",
                        "type": "Identifier",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": true,
                      "type": "Property",
                      "value": {
                        "name": "x",
                        "type": "Identifier",
                      },
                    },
                    {
                      "computed": false,
                      "key": {
                        "name": "y",
                        "type": "Identifier",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": true,
                      "type": "Property",
                      "value": {
                        "name": "y",
                        "type": "Identifier",
                      },
                    },
                    {
                      "computed": false,
                      "key": {
                        "name": "z",
                        "type": "Identifier",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": true,
                      "type": "Property",
                      "value": {
                        "name": "z",
                        "type": "Identifier",
                      },
                    },
                  ],
                  "type": "ObjectPattern",
                },
                "right": {
                  "properties": [
                    {
                      "computed": false,
                      "key": {
                        "name": "x",
                        "type": "Identifier",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": 44,
                      },
                    },
                    {
                      "computed": false,
                      "key": {
                        "name": "y",
                        "type": "Identifier",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": 55,
                      },
                    },
                    {
                      "computed": false,
                      "key": {
                        "name": "z",
                        "type": "Identifier",
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": false,
                      "type": "Property",
                      "value": {
                        "type": "Literal",
                        "value": 66,
                      },
                    },
                  ],
                  "type": "ObjectExpression",
                },
                "type": "AssignmentPattern",
              },
            ],
            "type": "ArrayPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* h([cls = class {}, xCls = class X {}, xCls2 = class { static name() {} }]) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": {
          "name": "h",
          "type": "Identifier",
        },
        "params": [
          {
            "elements": [
              {
                "left": {
                  "name": "cls",
                  "type": "Identifier",
                },
                "right": {
                  "body": {
                    "body": [],
                    "type": "ClassBody",
                  },
                  "id": null,
                  "superClass": null,
                  "type": "ClassExpression",
                },
                "type": "AssignmentPattern",
              },
              {
                "left": {
                  "name": "xCls",
                  "type": "Identifier",
                },
                "right": {
                  "body": {
                    "body": [],
                    "type": "ClassBody",
                  },
                  "id": {
                    "name": "X",
                    "type": "Identifier",
                  },
                  "superClass": null,
                  "type": "ClassExpression",
                },
                "type": "AssignmentPattern",
              },
              {
                "left": {
                  "name": "xCls2",
                  "type": "Identifier",
                },
                "right": {
                  "body": {
                    "body": [
                      {
                        "computed": false,
                        "key": {
                          "name": "name",
                          "type": "Identifier",
                        },
                        "kind": "method",
                        "static": true,
                        "type": "MethodDefinition",
                        "value": {
                          "async": false,
                          "body": {
                            "body": [],
                            "type": "BlockStatement",
                          },
                          "generator": false,
                          "id": null,
                          "params": [],
                          "type": "FunctionExpression",
                        },
                      },
                    ],
                    "type": "ClassBody",
                  },
                  "id": null,
                  "superClass": null,
                  "type": "ClassExpression",
                },
                "type": "AssignmentPattern",
              },
            ],
            "type": "ArrayPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* h([fn = function () {}, xFn = function x() {}] = []) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": {
          "name": "h",
          "type": "Identifier",
        },
        "params": [
          {
            "left": {
              "elements": [
                {
                  "left": {
                    "name": "fn",
                    "type": "Identifier",
                  },
                  "right": {
                    "async": false,
                    "body": {
                      "body": [],
                      "type": "BlockStatement",
                    },
                    "generator": false,
                    "id": null,
                    "params": [],
                    "type": "FunctionExpression",
                  },
                  "type": "AssignmentPattern",
                },
                {
                  "left": {
                    "name": "xFn",
                    "type": "Identifier",
                  },
                  "right": {
                    "async": false,
                    "body": {
                      "body": [],
                      "type": "BlockStatement",
                    },
                    "generator": false,
                    "id": {
                      "name": "x",
                      "type": "Identifier",
                    },
                    "params": [],
                    "type": "FunctionExpression",
                  },
                  "type": "AssignmentPattern",
                },
              ],
              "type": "ArrayPattern",
            },
            "right": {
              "elements": [],
              "type": "ArrayExpression",
            },
            "type": "AssignmentPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* yield() { }); 1`] = `
"SyntaxError [1:17-1:22]: 'yield' may not be used as an identifier in this context
> 1 | (async function* yield() { });
    |                  ^^^^^ 'yield' may not be used as an identifier in this context"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function* yield() { }); 2`] = `
"SyntaxError [1:17-1:22]: 'yield' may not be used as an identifier in this context
> 1 | (async function* yield() { });
    |                  ^^^^^ 'yield' may not be used as an identifier in this context"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*() { } = 1); 1`] = `
"SyntaxError [1:23-1:24]: Invalid left-hand side in assignment
> 1 | (async function*() { } = 1);
    |                        ^ Invalid left-hand side in assignment"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*() { await: 1; }); 1`] = `
"SyntaxError [1:21-1:26]: cannot use "await" as identifier inside an async function
> 1 | (async function*() { await: 1; });
    |                      ^^^^^ cannot use "await" as identifier inside an async function"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*([...x, y] = [1, 2, 3]) {}) 1`] = `
"SyntaxError [1:27-1:28]: The binding declaration is not destructible
> 1 | (async function*([...x, y] = [1, 2, 3]) {})
    |                            ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*({ w: { x, y, z } = { x: 4, y: 5, z: 6 } } = { w: undefined }) { }) 1`] = `
{
  "body": [
    {
      "expression": {
        "async": true,
        "body": {
          "body": [],
          "type": "BlockStatement",
        },
        "generator": true,
        "id": null,
        "params": [
          {
            "left": {
              "properties": [
                {
                  "computed": false,
                  "key": {
                    "name": "w",
                    "type": "Identifier",
                  },
                  "kind": "init",
                  "method": false,
                  "shorthand": false,
                  "type": "Property",
                  "value": {
                    "left": {
                      "properties": [
                        {
                          "computed": false,
                          "key": {
                            "name": "x",
                            "type": "Identifier",
                          },
                          "kind": "init",
                          "method": false,
                          "shorthand": true,
                          "type": "Property",
                          "value": {
                            "name": "x",
                            "type": "Identifier",
                          },
                        },
                        {
                          "computed": false,
                          "key": {
                            "name": "y",
                            "type": "Identifier",
                          },
                          "kind": "init",
                          "method": false,
                          "shorthand": true,
                          "type": "Property",
                          "value": {
                            "name": "y",
                            "type": "Identifier",
                          },
                        },
                        {
                          "computed": false,
                          "key": {
                            "name": "z",
                            "type": "Identifier",
                          },
                          "kind": "init",
                          "method": false,
                          "shorthand": true,
                          "type": "Property",
                          "value": {
                            "name": "z",
                            "type": "Identifier",
                          },
                        },
                      ],
                      "type": "ObjectPattern",
                    },
                    "right": {
                      "properties": [
                        {
                          "computed": false,
                          "key": {
                            "name": "x",
                            "type": "Identifier",
                          },
                          "kind": "init",
                          "method": false,
                          "shorthand": false,
                          "type": "Property",
                          "value": {
                            "type": "Literal",
                            "value": 4,
                          },
                        },
                        {
                          "computed": false,
                          "key": {
                            "name": "y",
                            "type": "Identifier",
                          },
                          "kind": "init",
                          "method": false,
                          "shorthand": false,
                          "type": "Property",
                          "value": {
                            "type": "Literal",
                            "value": 5,
                          },
                        },
                        {
                          "computed": false,
                          "key": {
                            "name": "z",
                            "type": "Identifier",
                          },
                          "kind": "init",
                          "method": false,
                          "shorthand": false,
                          "type": "Property",
                          "value": {
                            "type": "Literal",
                            "value": 6,
                          },
                        },
                      ],
                      "type": "ObjectExpression",
                    },
                    "type": "AssignmentPattern",
                  },
                },
              ],
              "type": "ObjectPattern",
            },
            "right": {
              "properties": [
                {
                  "computed": false,
                  "key": {
                    "name": "w",
                    "type": "Identifier",
                  },
                  "kind": "init",
                  "method": false,
                  "shorthand": false,
                  "type": "Property",
                  "value": {
                    "name": "undefined",
                    "type": "Identifier",
                  },
                },
              ],
              "type": "ObjectExpression",
            },
            "type": "AssignmentPattern",
          },
        ],
        "type": "FunctionExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*(a = super()) { }); 1`] = `
"SyntaxError [1:26-1:27]: Calls to super must be in the "constructor" method of a class expression or class declaration that has a superclass
> 1 | (async function*(a = super()) { });
    |                           ^ Calls to super must be in the "constructor" method of a class expression or class declaration that has a superclass"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*(a = super()) { }); 2`] = `
"SyntaxError [1:26-1:27]: Calls to super must be in the "constructor" method of a class expression or class declaration that has a superclass
> 1 | (async function*(a = super()) { });
    |                           ^ Calls to super must be in the "constructor" method of a class expression or class declaration that has a superclass"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async function*(x = await 1) { }); 1`] = `
"SyntaxError [1:21-1:26]: Await expression not allowed in formal parameter
> 1 | (async function*(x = await 1) { });
    |                      ^^^^^ Await expression not allowed in formal parameter"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > (async.foo6 => 1) 1`] = `
"SyntaxError [1:12-1:14]: Expected ')'
> 1 | (async.foo6 => 1)
    |             ^^ Expected ')'"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > 0, async function* g(...a,) {}; 1`] = `
"SyntaxError [1:25-1:26]: The binding declaration is not destructible
> 1 | 0, async function* g(...a,) {};
    |                          ^ The binding declaration is not destructible"
`;

exports[`Expressions - Async Generator > Expressions - Async Generator (pass) > 0, async function* g(...x = []) {} 1`] = `
"SyntaxError [1:30-1:31]: The binding declaration is not destructible
> 1 | 0, async function* g(...x = []) {}
    |                               ^ The binding declaration is not destructible"
`;
